### ISSUE TITLE
Per-app custom settings API

### ISSUE THREAD

**Issue by eloquence** [mastodon/mastodon#28265]
### Pitch

To empower app developers, add an API endpoint to enable apps that have been granted write access to the API to also store JSON configuration objects of a limited size (e.g., 32 KB per app).

This will lead to more innovation in the Mastodon app ecosystem beyond configuration options that Mastodon has already thought of.

### Motivation

Third party apps may offer preferences that go beyond that the Mastodon official app offers. For example, Phanpy lets the user browse timelines on multiple instances. 

If an app developer wants to store and synchronize such app-specific settings, they have two primary options:

- run their own configuration service (creating a single point of failure)
- allow the user to export and import app-specific settings manually (this is what Phanpy currently does)

The first option is user-friendly but contradicts the fediverse's overall design goals; the second option is not user-friendly.

An endpoint to set and retrieve per-app settings significantly expands the possibility space of what apps can do, without being tied to a centralized service.

## Security considerations

- Any app should only be able to set and retrieve its own settings, not any other app's settings.
- Any endpoint that writes private date may raise eyebrows; however, there are already other ways in which an app could potentially write private data to a server in ways that are difficult to detect, e.g., by adding private notes to an account. Limiting size of such configuration blobs offers mitigation against significant misuse without compromising user privacy. However, it may be desirable to offer the user a quick way in the web UI to access any app's settings data.

**Comment by connyduck** [comment]
> Any app should only be able to set and retrieve its own settings, not any other app's settings.

I don't think that is possible. How would a server know it is the same app again and not an imposter?
I'd rather have all custom settings read and writeable by all apps so if multiple apps support a feature they can sync as well. 
An app that messes up settings of other apps would probably lose it reputation pretty fast, so it shouldn't be a problem.

**Comment by andypiper** [comment]
I'd have to agree that this is not something that is possible, beyond just the app and user security. In particular if we think about this:

> Third party apps may offer preferences that go beyond that the Mastodon official app offers. For example, Phanpy lets the user browse timelines on multiple instances.

In this case, where you're talking about multiple instances (fwiw, I love this feature in Ice Cubes as well), the app would still need to connect to a specific instance first to get the stored information, and you're back to the single point of failure.

I don't think this is likely to be a high priority for the moment. Something like a user-authenticated shared storage system (iCloud or Firebase are "obvious" candidates on mobile, but are not FLOSS options) would probably be better options, rather than building it in to the Mastodon platform.

**Comment by strk** [comment]
> How would a server know it is the same app again and not an imposter?

In the same way it grants it the ability to write posts on the behalf of the user ?

**Comment by connyduck** [comment]
> > How would a server know it is the same app again and not an imposter?
> 
> In the same way it grants it the ability to write posts on the behalf of the user ?

It grants this ability to all apps, not only to one specific app. That is a small but important difference. 
According to the feature request app A should be able to save preferences and then read them again from another device, while app B can't access them. So I'm asking what should stop app B from telling the server "hey I'm app A give me the preferences I just saved"?

**Comment by strk** [comment]
On Wed, Dec 13, 2023 at 09:07:40AM -0800, Konrad Pozniak wrote:
> It grants this ability to all apps, not only to one specific app. That is a small but important difference. 
> According to the feature request app A should be able to save preferences and then read them again from another device, while app B can't access them. So I'm asking what should stop app B from telling the server "hey I'm app A give me the preferences I just saved"?

I never developed an OAuth application myself BUT I'm used to see that
the authorization token is allowed to do different operations, so
there could be an operation being "write and read configuration X"

Now, "configuration X" could be a generic object or a per-app
configuration. If it is a per-app configuration there should be
another operation that is "register new per-app configuration".

**Comment by ThisIsMissEm** [comment]
I think this would be possible with the following:

- A `write:app:storage` and `read:app:storage` OAuth 2.0 scope
- Have the stored data be bound to the OAuth 2.0 Application (this means apps that don't remember their client_id & client_secret per server couldn't use this), since you'd have "different apps"
- Limit storage to 1-2kb, and absolutely no more, since this could certainly be used as an attack surface for resource exhaustion.

Before this could be implemented we'd need #27478 implemented, such that we could give insight for server admins to understand which apps are using resources on their servers.

**Comment by cyrneko** [comment]
> Any app should only be able to set and retrieve its own settings, not any other app's settings.

I feel this is very difficult, if not impossible to do in practice. Perhaps such a system should instead be written and documented with the implication that any of this data is public to any client that authenticated with a given account, to encourage developers to not put private data into there in the first place.
